File: messagematch/src/main/java/org/forwoods/messagematch/generate/JsonGenerator.java
package org.forwoods.messagematch.generate;

import java.io.IOException;
import java.io.InputStream;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Iterator;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.stream.Collectors;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import org.forwoods.messagematch.generate.nodegenerators.*;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.forwoods.messagematch.generate.nodegenerators.constraints.ProvidedConstraint;
import org.forwoods.messagematch.matchgrammar.MatcherLexer;
import org.forwoods.messagematch.matchgrammar.MatcherParser;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ValueNode;
import org.forwoods.messagematch.util.PathExtractor;

import static org.forwoods.messagematch.match.JsonMatcher.sizePattern;

/**
 * @author Tom
 *
 */
public class JsonGenerator {
	static final ObjectMapper mapper = new ObjectMapper();

	private final JsonNode matcherNode;

	private final Map<String, ValueProvider> bindings;

	private final Map<String, Object> rawBindings;
	
	protected long genTime=-1;//can be overridden for unit tests of the generator

	public JsonGenerator(InputStream matcherFile, Map<String, Object> bindings) throws IOException {
		this(mapper.readTree(matcherFile), bindings);
	}
	public JsonGenerator(InputStream matcherFile) throws IOException {
		this(mapper.readTree(matcherFile), Map.of());
	}

	public JsonGenerator(String matcherString, Map<String, Object> bindings) throws JsonProcessingException {
		this(mapper.readTree(matcherString), bindings);
	}
	public JsonGenerator(String matcherString) throws IOException {
		this(mapper.readTree(matcherString), Map.of());
	}
	public JsonGenerator(JsonNode matcherNode) {
		this(matcherNode, Map.of());
	}
	public JsonGenerator(JsonNode matcherNode, Map<String, Object> bindings) {
		this.matcherNode = matcherNode;
		this.bindings=bindings.entrySet().stream()
				.collect(
						Collectors.toMap(Map.Entry::getKey, e-> createValueProvider(e.getValue().toString())));
		this.rawBindings = bindings;
	}

	public static ValueProvider createValueProvider(String val) {
		ValueProvider res = new ValueProvider();
		res.addConstraint(new ProvidedConstraint(val));
		return res;
	}

	public JsonNode generate() {

		if (genTime<0) genTime = System.currentTimeMillis();
		Instant i = Instant.ofEpochMilli(genTime);
		ZonedDateTime t = ZonedDateTime.ofInstant(i, ZoneOffset.UTC);
		LocalDate d = t.toLocalDate();
		LocalTime localTime = t.toLocalTime();
		bindings.put("date", createValueProvider(d.toString()));
		bindings.put("time", createValueProvider(localTime.toString()));
		bindings.put("datetime", createValueProvider(Long.toString(genTime)));
		
		return generate(matcherNode).generate();
	}

	public String generateString() {
		return generate().toPrettyString();
	}

	public NodeGenerator generate(JsonNode matcherNode) {
		switch (matcherNode.getNodeType()) {
		case STRING:
		case BOOLEAN:
		case NUMBER:
			return generatePrimitive((ValueNode) matcherNode);

		case OBJECT:
			return generateObject((ObjectNode) matcherNode);
		case POJO:
		case MISSING:
		case BINARY:
			break;
		case ARRAY:
			return generateArray((ArrayNode) matcherNode);
		case NULL:
			break;
		default:
			break;

		}

		return new LiteralGenerator(matcherNode);

	}

	private NodeGenerator generatePrimitive(ValueNode matcherNode) {
		String matcher = matcherNode.asText();
		if (matcher.startsWith("$ID")) {
			JsonNode bound = PathExtractor.extractPrimitiveNode(matcher,rawBindings);
			return new ReferenceGenerator(bound);
		}
		if (matcher.startsWith("$")) {
			return parseMatcher(matcher, bindings);
		}
		if (matcher.startsWith("\\$")) {
			return new LiteralGenerator(JsonNodeFactory.instance.textNode(matcher.replaceFirst("\\$", "$")));
		}
		return new LiteralGenerator(matcherNode);
	}
	
	public static NodeGenerator parseMatcher(String matcher, Map<String, ValueProvider> bindings) {
		MatcherLexer l = new MatcherLexer(CharStreams.fromString(matcher));
		MatcherParser p = new MatcherParser(new CommonTokenStream(l));
		p.addErrorListener(new BaseErrorListener() {
	        @Override
	        public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
	            throw new IllegalStateException("failed to parse at line " + line + " due to " + msg, e);
	        }
	    });
		GrammarListenerGenerator listener = new GrammarListenerGenerator(bindings);
		p.addParseListener(listener);
		p.matcher();
		return listener.result;
	}

	private NodeGenerator generateArray(ArrayNode matcherNode) {
		ArrayTypeGenerator arrayGen = new ArrayTypeGenerator();
		ArraySizeGenerator sizeGen = null;
		for (JsonNode jsonNode : matcherNode) {
			NodeGenerator generate = generate(jsonNode);
			if (generate instanceof ArraySizeGenerator) {
				sizeGen = (ArraySizeGenerator) generate;
				continue;
			}
			arrayGen.addChild(generate);
		}

		if (sizeGen!=null) {
			arrayGen.setSize(sizeGen.getMin(), sizeGen.getMax());
		}

		return arrayGen;
	}

	private NodeGenerator generateObject(ObjectNode matcherNode) {
		ObjectTypeGenerator object = new ObjectTypeGenerator();
		for (Iterator<Map.Entry<String, JsonNode>> iterator = matcherNode.fields(); iterator.hasNext();) {
			Map.Entry<String, JsonNode> child = iterator.next();
			String key = child.getKey();
			if (key.equals("$Strict")) continue;
			if (key.equals("$Size")) {
				JsonNode node = child.getValue();
				String bounds = node.asText();
				Matcher m = sizePattern.matcher(bounds);
				Integer min = null;
				Integer max = null;
				if (m.matches()) {
					if (m.group(1).length() > 0) {
						min = Integer.valueOf(m.group(1));
					}
					if (m.group(2).length() > 0) {
						max = Integer.valueOf(m.group(2));
					}
				}
				return new ArraySizeGenerator(min, max);}
			if (key.equals("$ID")) {
				JsonNode bound = PathExtractor.extractObjectNode(child.getValue().asText(),rawBindings);
				if (bound!=null)
					return new ReferenceGenerator(bound);
				continue;
			}
			if (key.startsWith("$")) {
				NodeGenerator gen = parseMatcher(key, bindings);
				JsonNode generate = gen.generate();
				key = generate.asText();
			}
			object.addChild(key, generate(child.getValue()));
		}
		return object;
	}

}

File: messagematch-typescript/src/generate/JsonGenerator.ts
import {Temporal} from '@js-temporal/polyfill';
import {ValueProvider} from './ValueProvider';
import {ProvidedConstraint} from './constraints/ProvidedConstraint';
import {NodeGenerator} from './nodegenerators/NodeGenerator';
import {LiteralGenerator} from './nodegenerators/LiteralGenerator';
import {ReferenceGenerator} from './nodegenerators/ReferenceGenerator';
import {PathExtractor} from './utils/PathExtractor';
import {ObjectTypeGenerator} from "./nodegenerators/ObjectTypeGenerator";
import MatcherLexer from "../antlr4ts/org/forwoods/messagematch/matchgrammar/MatcherLexer";
import MatcherParser from "../antlr4ts/org/forwoods/messagematch/matchgrammar/MatcherParser";
import {GrammarListenerGenerator} from "./GrammarListenerGenerator";
import {ArraySizeGenerator, ArrayTypeGenerator} from "./nodegenerators/ArrayTypeGenerator";

export class JsonGenerator {
    // genTime can be set by tests to make generation deterministic
    genTime: number = -1;
    private matcherNode: any;
    private bindings: Record<string, ValueProvider>;
    private rawBindings: Record<string, any>;
    static readonly mapper = {}; // stub for now

    constructor(matcher: any, bindings: Record<string, any> = {}) {
        this.matcherNode = matcher;
        this.rawBindings = bindings;
        this.bindings = Object.fromEntries(
            Object.entries(bindings).map(([k, v]) => [k, this.createValueProvider(v.toString())])
        );
    }

    private parseMatcher(matcher: string): NodeGenerator {
        const antlr4 = require('antlr4');
        const chars = new antlr4.InputStream(matcher);
        const lexer = new MatcherLexer(chars);
        const tokens = new antlr4.CommonTokenStream(lexer);
        const parser = new MatcherParser(tokens);
        // throw on syntax errors
        parser.removeErrorListeners();
        parser.addErrorListener({
            syntaxError(_recognizer:any, _offendingSymbol:any, line:number, _charPos:number, msg:string) {
                throw new Error(`failed to parse at line ${line} due to ${msg}`);
            }
        });
        const grammarListenerGenerator = new GrammarListenerGenerator(this.bindings);
        parser.addParseListener(grammarListenerGenerator);
        parser.matcher();
        if (!grammarListenerGenerator.result) {
            throw new Error("Parsing of grammar failed");
        }
        return grammarListenerGenerator.result;
    }

    generate(): any {
        if (this.genTime < 0) {
            this.genTime = Temporal.Now.instant().epochMilliseconds;
        }

        const instant = Temporal.Instant.fromEpochMilliseconds(this.genTime);
        const zonedDateTime = instant.toZonedDateTimeISO('UTC');
        const date = zonedDateTime.toPlainDate();
        const time = zonedDateTime.toPlainTime();

        this.bindings['date'] = this.createValueProvider(date.toString());
        this.bindings['time'] = this.createValueProvider(time.toString());
        this.bindings['datetime'] = this.createValueProvider(this.genTime.toString());

        return this.generateNode(this.matcherNode).generate();
    }

    private generateNode(matcherNode: any): NodeGenerator {
        if (typeof matcherNode === 'string' || typeof matcherNode === 'boolean' || typeof matcherNode === 'number') {
            return this.generatePrimitive(matcherNode);
        }

        if (Array.isArray(matcherNode)) {
            return this.generateArray(matcherNode);
        }

        if (matcherNode === null) {
            return this.generateLiteral(null);
        }

        if (typeof matcherNode === 'object') {
            return this.generateObject(matcherNode);
        }

        return this.generateLiteral(matcherNode);
    }

    private generatePrimitive(matcherNode: any): NodeGenerator {
        const matcher = typeof matcherNode === 'object' ? matcherNode.asText?.() || matcherNode.toString() : matcherNode.toString();

        if (matcher.startsWith('$ID')) {
            const bound = PathExtractor.extractPrimitiveNode(matcher, this.rawBindings);
            return new ReferenceGenerator(bound);
        }

        if (matcher.startsWith('$')) {
            return this.parseMatcher(matcher);
        }

        if (matcher.startsWith('\\$')) {
            return new LiteralGenerator(matcher.replace(/^\\\$/, '$'));
        }

        return new LiteralGenerator(matcherNode);
    }

    private generateArray(arr: any[]): NodeGenerator {
        const arrayGen = new ArrayTypeGenerator();
        let sizeGen:ArraySizeGenerator | undefined
        arr.forEach(item => {
            const generate:NodeGenerator = this.generateNode(item);
            if (generate instanceof ArraySizeGenerator) {
               sizeGen = generate as ArraySizeGenerator;
            } else {
                arrayGen.addChild(generate);
            }
        });
        if (!!sizeGen) {
            arrayGen.setSize(sizeGen.min, sizeGen.max);
        }
        return arrayGen;
    }

    private sizePattern = /([0-9]*)-([0-9]*)/;
    private generateObject(matcherNode: any): NodeGenerator {
        const objectTypeGenerator:any = new ObjectTypeGenerator();
        let ownPropertyNames:string[] = Object.getOwnPropertyNames(matcherNode);
        for (let k of ownPropertyNames) {
            let value = matcherNode[k];
            if (k == "$Strict") continue;
            if (k == "$Size") {
                const bounds: string = value;
                const m = this.sizePattern.exec(bounds);
                if (m) {
                    const min = m[1] && m[1].length > 0 ? Number(m[1]) : undefined;
                    const max = m[2] && m[2].length > 0 ? Number(m[2]) : undefined;
                    return new ArraySizeGenerator(min, max);
                }
                return new ArraySizeGenerator();
            }
            if (k == "$ID") continue; //TODO implement this
            if (k.startsWith("$")) {
                const gen = this.parseMatcher(k);
                k = gen.generate();
            }
            objectTypeGenerator.addChild(k, this.generateNode(value));
        }
        return objectTypeGenerator;
    }

    private generateLiteral(value: any): NodeGenerator {
        return new LiteralGenerator(value);
    }

    private createValueProvider(val: string): ValueProvider {
        const res = new ValueProvider();
        res.addConstraint(new ProvidedConstraint(val));
        return res;
    }

    getBindings(): Record<string, string> {
        const result:Record<string, string> = {};
        Object.entries(this.bindings).forEach(([key, value]) => {
            result[key] = value.asString();
        });
        return result;
    }
}

export default JsonGenerator;
